packages/playwright/src/util.ts

```diff
@@ -319,14 +319,17 @@ export function resolveImportSpecifierExtension(resolved: string): string | unde
     break;  // Do not try '' when a more specific extension like '.jsx' matched.
   }
 
-  if (dirExists(resolved)) {
+  // Following TypeScript's path mapping logic, index files and package.json are not resolved in ESM.
+  // TypeScript does not interpret package.json for path mappings: https://www.typescriptlang.org/docs/handbook/modules/reference.html#paths-should-not-point-to-monorepo-packages-or-node_modules-packages
+  const shouldNotResolveDirectory = isPathMapping && isESM;
+
+  if (!shouldNotResolveDirectory && dirExists(resolved)) {
```

Comment:
It seems like TypeScript:
- does not honor `exports` field from `package.json` after the type mapping;
- does honor the default `index.js` and `main` field from `package.json` after the type mapping as explained here: https://www.typescriptlang.org/docs/handbook/modules/reference.html#directory-modules-index-file-resolution. Note that both `index.js` and `main` are ignored for ESM, as explained here: https://nodejs.org/dist/latest-v20.x/docs/api/modules.html#folders-as-modules.

The docs above match my manual testing as well. I think we should follow TS and implement the same logic.

Reply:
I've added a test to ensure we follow your first observation in https://github.com/microsoft/playwright/pull/32078/commits/9a80386c7d057027499609aa4d648549482474fa.

For the second one, isn't that exactly the behaviour we have right now? If there's a `package.json` file, then we let Node.js deal with resolving to either `index.js` or `main`, depending on the ESM mode.

Reply:
- For the `exports` field, it looks like the code will defer to Node.js (when not in ESM), and Node will honor the `exports`. Is that not the case?
- For the `main` field, TypeScript also checks various extensions (e.g. `ts`) after following it, and we currently do not, because we leave it to Node.js. That's a minor thing, but it would be ideal to mirror TypeScript there as well.
- Updating comments to explain these things would be great as well!

Reply:
1. I believe the test https://github.com/Skn0tt/playwright/blob/9a80386c7d057027499609aa4d648549482474fa/tests/playwright-test/resolver.spec.ts#L609-L644 asserts just that behaviour. Maybe i'm looking at it wrong though.
2. The `.ts` extension is covered by https://github.com/Skn0tt/playwright/blob/9a80386c7d057027499609aa4d648549482474fa/tests/playwright-test/resolver.spec.ts#L572-L607. What other extensions should we cover?

Both of these are about the path mapping case. Are you referring to the case where there's no path mapping?

In https://github.com/microsoft/playwright/pull/32078/commits/62ff3fd736429eb9f3e4038694e4d192944489f8, i've added a test to ensure that `main` is ignored in ESM mode, and updated the comments.

Reply:
Yep, the tests look like they cover all the cases. I guess for the item 1, Node.js does not look at `exports` either, which works for us. Thank you for following through!